PARQUET-2081: Encryption translation tool - Parquet-hadoop

shangxinli opened the pull request:
Summary:
Design doc - High Throughput CLAC Writer: https://docs.google.com/document/d/1-XdE8-QyDHnBsYrClwNsR8X3ks0JmKJ1-rXq7_th0hc

Added unit tests

Integration tests with real data

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

shangxinli commented:
@ggershinsky @gszadovszky Can you have a look?

ggershinsky commented:
Hi @shangxinli , will be glad to. We got a number of national holidays here, but I'll get to this soon.

gszadovszky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+    return prunePaths;
+  }
+
+  public enum EncryptMode {
```
Why don't we use the `ParquetCipher` enum  directly?

shangxinli replied:
Yes, we should.

gszadovszky commented on pom.xml:
```diff
@@ -522,6 +522,7 @@
                 change to fix a integer overflow issue.
                 TODO: remove this after Parquet 1.13 release -->
               <exclude>org.apache.parquet.column.values.dictionary.DictionaryValuesWriter#dictionaryByteSize</exclude>
+              <exclude>org.apache.parquet.hadoop.util</exclude>
```
With this exclusion we won't check API compatibility for this package. If want to do this we should state somewhere that we don't guarantee backward compatibility of the classes in this package. We've had bad times doing things like this. Don't we have other solutions to keep backward compatibility instead?

shangxinli replied:
Let me narrow it down to the impacted classes only.

gszadovszky replied:
Maybe I was misleading. For me it doesn't matter if we have a package or classes listed here. My problem is that we are breaking backward compatibility from the API point of view. If we think it is OK to break the compatibility of this package then I am fine having it here but with a separate comment. (For `DictionaryValuesWriter` we did not have a choice to not to break the API that's why the TODO is there.)
So, if the `util` package is not for public use than we shall add `@Private` to all of its classes and maybe also add a `package-info.java` stating that this is not for public use and add similar comments here as well. Otherwise, we should maintain backward compatibility for these classes as well.

gszadovszky reviewed (changes requested):
I think the concept is good and the implementation seems also fine. I have some comments, though.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+import static org.apache.parquet.hadoop.ParquetWriter.DEFAULT_BLOCK_SIZE;
+import static org.apache.parquet.hadoop.ParquetWriter.MAX_PADDING_SIZE_DEFAULT;
+
+public class ColumnEncryptor {
```
sure

gszadovszky replied:
Please add `static` as well.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+    this.conf = conf;
+  }
+
+  /**
+   *
+   * @param inputFile Input file
+   * @param outputFile Output file
+   * @throws IOException
+   */
```
fixed

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+    }
+  }
+
+  private final int pageBufferSize = ParquetProperties.DEFAULT_PAGE_SIZE * 2;
```
yeah, fixed

gszadovszky replied:
Please add `static` as well.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+      for (int i = 0; i < columnsInOrder.size(); i += 1) {
+        ColumnChunkMetaData chunk = columnsInOrder.get(i);
+        ColumnDescriptor descriptor = descriptorsMap.get(chunk.getPath());
+        processChunk(descriptor, chunk, reader, writer, nullifyColumns, blockId, i, meta.getFileMetaData().getCreatedBy());
```
Great idea! I am thinking if we introduce the universal tool, the file could be rewritten multiple times. How about "original.writer.version" and "rewriter. version", and the "rewriter. version" can be comma-separated and we can append to for every time rewriting. Let me know your thinking and I can implement it once we agree.

gszadovszky replied:
Actually, I've suggested only one additional field and the other would be the already existing `created_by`. The question is which one should we add to `created_by` and which one to the key-value metadata in addition. Since we do not decode the data (so we cannot validate it) I think it is good to copy the `created_by` field from the original file and add the "rewriter version" to the key-value metadata. But it is a tricky question. Let's say we have a bug in one of the compression codecs so we generate invalid compressed pages with a certain version or parquet-mr. Now, we have a parquet file that we "trans-compress" the pages with a tool with this certain parquet-mr version and codec. Then, we use another tool to encrypt the file. I think this topic is more for parquet-format and the whole Parquet community. We shall introduce the discussed and accepted solution for any tool that "rewrites" a file while keeping the original data "undecoded".

Some questions (about a common rewriter) we need to answer before step forward:
* What would be the place of the rewriter versions? (New specific field or key-value metadata? Which key shall we use?)
* Shall we somehow also save what the rewriter has done? How?
* At what level shall we copy the original `created_by` field and what level shall we write the version of the rewriter to that field instead? (What different levels are possible?)
* From the introduction of this rewriter(s) field in case of any related writer version dependent fix we need to check this field as well and not only the `created_by` one.

shangxinli replied:
These are all great questions! I think we can start the conversation with some proposals and let the community review them. I just created the doc for the 'Unified Apache Rewriter' https://docs.google.com/document/d/1Ryt5uXnp-YwOrsnIDrGdTMoFfbOaTM6X39pBLHOa_50 and put all the questions into the open questions.

gszadovszky replied:
I agree it is a bigger question. It is completely fine to handle separately from this PR. Meanwhile, we should ensure this won't get released before having the new field filled correctly. I would suggest having this tracked in a jira that is targeted for 1.13.0.

shangxinli replied:
https://issues.apache.org/jira/browse/PARQUET-2093 is created

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/ColumnChunkMetaData.java:
```diff
@@ -370,6 +370,14 @@ public boolean hasDictionaryPage() {
+   */
+  @Private
+  public boolean isEncrypted() {
+    return this instanceof EncryptedColumnChunkMetaData;
```
Make sense!

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,351 @@
+    }
+  }
+
+  private byte[] encryptPageLoad(TransParquetFileReader reader, int payloadLength, BlockCipher.Encryptor dataEncryptor, byte[] AAD) throws IOException {
```
Can you clarify a little bit? Do you mean changing the method name? Or refactor the code?

gszadovszky replied:
Sorry, I've meant the method name `encryptPageLoad` feel strange to me. I think it is either payload or page.

ggershinsky commented:
Significant functionality and good implementation. 
Here goes the first round of my comments. Once our holidays are over (week of 9/27), I'll continue to work on this.

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/hadoop/util/EncDecProperties.java:
```diff
@@ -0,0 +1,96 @@
+    return FileDecryptionProperties.builder().withPlaintextFilesAllowed().withoutFooterSignatureVerification().withKeyRetriever(keyRetriever).build();
+  }
+
+  public static FileDecryptionProperties getFileDecryptionPropertiesNonExistKey() throws IOException {
```
this function is not used (?)

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java:
```diff
@@ -599,6 +599,31 @@ private void innerWriteDataPage(
+   * @param dlEncoding encoding of the definition level
+   * @param valuesEncoding encoding of values
+   * @param blockEncryptor encryptor for block data
+   * @param AAD AAD
```
I just followed the existing name, but this is correct feedback. So I changed not only this one but also several other places to rename AAD to pageHeadAAD.

shangxinli commented:
@ggershinsky @gszadovszky Just pushed another commit to address more feedback. Can you have another look?

gszadovszky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,314 @@
+    return BytesInput.from(data, 0, length);
+  }
+
+  private int toIntWithCheck(long size) {
```
We already have `Math.toIntExact(long)` to do a safe downcast.

gszadovszky commented on pom.xml:
```diff
@@ -522,6 +522,9 @@
                 change to fix a integer overflow issue.
                 TODO: remove this after Parquet 1.13 release -->
               <exclude>org.apache.parquet.column.values.dictionary.DictionaryValuesWriter#dictionaryByteSize</exclude>
+              <exclude>org.apache.parquet.hadoop.util.ColumnMasker.MaskMode</exclude>
+              <exclude>org.apache.parquet.hadoop.util.CompressionConverter</exclude>
+              <exclude>org.apache.parquet.hadoop.util.ColumnMasker</exclude>
```
You've accepted my previous comment but nothing changed.

> (...) if the `util` package is not for public use than we shall add `@Private` to all of its classes and maybe also add a `package-info.java` stating that this is not for public use and add similar comments here as well. Otherwise, we should maintain backward compatibility for these classes as well.

shangxinli replied:
@gszadovszky, Sorry my bad. Making it private won't work because it is still used outside util package. So I just revert the refactoring code to avoid the exclusion here. We can do it later.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:
```diff
@@ -503,7 +503,7 @@ private void addRowGroup(ParquetMetadata parquetMetadata, List<RowGroup> rowGrou
       if (null != fileEncryptor) {
         columnOrdinal++;
-        columnSetup = fileEncryptor.getColumnSetup(path, false, columnOrdinal);
+        columnSetup = fileEncryptor.getColumnSetup(path, true, columnOrdinal);
```
yep, it's ok to call getColumnSetup for every column. In any case, the true/false here is not an indication of whether column is encrypted or not. This boolean is simply a flag whether the column setup should have already been created at this point, a simple flow check.

ggershinsky replied:
this flag has to be false. If this triggers an exception, then there might be a problem.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,339 @@
+   * @param fileEncryptionProperties FileEncryptionProperties of the file
+   * @throws IOException
+   */
+  public void encryptColumns(String inputFile, String outputFile, List<String> paths, FileEncryptionProperties fileEncryptionProperties) throws IOException {
```
an excellent point. I agree it would be good to support this scenario. It is somewhat harder to explain and define (check if each column is already encrypted; if yes, compare its encryption properties). It might be easier to start with a simpler scenario, where the input files are not encrypted at all. But I'm ok either way.

ggershinsky replied:
please check if the file (or column) is already encrypted.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,307 @@
+  }
+
+  private void processBlocks(TransParquetFileReader reader, ParquetFileWriter writer, ParquetMetadata meta,
+                            MessageType schema, List<String> paths) throws IOException {
```
renamed from paths to encryptPaths and encrypPath to encryptColumnsPath.

gszadovszky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,339 @@
+    } else {
+      BloomFilter bloomFilter = reader.readBloomFilter(chunk);
+      ColumnIndex columnIndex = reader.readColumnIndex(chunk);
+      OffsetIndex offsetIndex = reader.readOffsetIndex(chunk);
```
Great catch, @ggershinsky and thanks for the heads up. Offset index contains the absolute offset and the size of each page so definitely shall be updated.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,307 @@
+      writer.startColumn(descriptor, chunk.getValueCount(), chunk.getCodec());
+      encryptPages(reader, chunk, writer, createdBy, blockId, columnId);
+      writer.endColumn();
+    } else {
```
@shangxinli per the discussion yesterday, just a thought. What if instead of this `else` block, the same sequence of
```  
      writer.startColumn(descriptor, chunk.getValueCount(), chunk.getCodec());
      encryptPages(reader, chunk, writer, createdBy, blockId, columnId);
      writer.endColumn();
```
is called for unencrypted columns? Of course, `encryptPages` would need to be renamed to something like `copyAndEncryptPages`, and should skip encryption for column that should not be encrypted. But this approach might be able to automatically re-create the column indexes for unencrypted columns, instead of reading and shifting them (in case the latter is challenging).

shangxinli replied:
I like the idea to streamline the code path. Only thing is some downgrade of the performance. I see a somewhat 10~20% downgrade comparing with calling appendColumnChunk directly. But I think comparing the risk,  it is worthy of doing that.

gszadovszky reviewed (changes requested):
Some unit tests are necessary to validate the filtering. (I would suggest validating the unit test by executing it with the previous implementation where we know offset indexes are wrong.)

Alternatively to using the same code path for encrypted/non-encrypted column chunks we might keep the separated paths by recalculating the offset index only. We only need to track the  difference of the offsets of the column chunk in the original offset index and the new file being written. (It would require to extend `OffsetIndex`.)

shangxinli commented:
@gszadovszky The first two commits had the OffsetIndex issue. The 3rd commit added test verifyOffsetIndexes() which verified not only the offset but also the page(header and payload) read with that offset index. That tests can reproduce the issue and solved the issue in the 3rd commit.  

The last commit is not solving any issue but just to take an alternative way(we think it is less error-prone). So no tests were added because the tests added earlier already covered that.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -280,8 +288,12 @@ private void encryptPages(TransParquetFileReader reader, ColumnChunkMetaData chu
   }
 
-  private byte[] encryptPayload(TransParquetFileReader reader, int payloadLength, BlockCipher.Encryptor dataEncryptor, byte[] AAD) throws IOException {
+  private byte[] encryptPayload(TransParquetFileReader reader, int payloadLength, BlockCipher.Encryptor dataEncryptor,
```
I will fix

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -180,21 +192,13 @@ private void processBlocks(TransParquetFileReader reader, ParquetFileWriter writ
 
-  private void encryptPages(TransParquetFileReader reader, ColumnChunkMetaData chunk,
-                             ParquetFileWriter writer, String createdBy, int blockId, int columnId) throws IOException {
+  private void encryptPages(TransParquetFileReader reader, ColumnChunkMetaData chunk, ParquetFileWriter writer,
```
I will fix

gszadovszky commented:
Sorry, @shangxinli. My bad. Ignore my comment about the testing.

shangxinli commented:
> Sorry, @shangxinli. My bad. Ignore my comment about the testing.

No problem Gabor! Thank you for all the feedbacks!

shangxinli commented:
@ggershinsky Do you want to have another look?

ggershinsky commented:
yep, a couple of quick comments. will comb again tomorrow.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java:
```diff
@@ -1125,7 +1154,7 @@ private static void serializeColumnIndexes(
         byte[] columnIndexAAD = null;
         if (null != fileEncryptor) {
-          InternalColumnEncryptionSetup columnEncryptionSetup = fileEncryptor.getColumnSetup(column.getPath(), false, cIndex);
+          InternalColumnEncryptionSetup columnEncryptionSetup = fileEncryptor.getColumnSetup(column.getPath(), true, cIndex);
```
this flag has to be false. If this triggers an exception, then there might be a problem.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,317 @@
+  private void processPages(TransParquetFileReader reader, ColumnChunkMetaData chunk, ParquetFileWriter writer,
+                            String createdBy, int blockId, int columnId, boolean encrypt) throws IOException {
+    int pageOrdinal = 0;
+    EncryptorRunTime encryptorRunTime = new EncryptorRunTime(writer.getEncryptor(), chunk, blockId, columnId);
```
not needed if `encrypt` is false

shangxinli replied:
Just create empty EncryptorRunTime when it is not 'encrypt'. The empty one will return 'null' for it's field members and the code below can handle nulls. This would avoid a lot of places 'if (encrypt)' that would make the code very messy.

ggershinsky replied:
agreed.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,317 @@
+          }
+          //No quickUpdatePageAAD needed for dictionary page
+          DictionaryPageHeader dictPageHeader = pageHeader.dictionary_page_header;
+          pageLoad = processPayload(reader, pageHeader.getCompressed_page_size(), encryptorRunTime.getDataEncryptor(), encryptorRunTime.getDictPageAAD(), encrypt);
```
should be handled differently if `encrypt` is false

shangxinli replied:
Inside processPayload() we treat the differently. That is why Gabor had earlier comment to change the method name from encryptPayload() to processPayload().

ggershinsky replied:
My issue was with the input parameters like `encryptorRunTime.getDataEncryptor()`. But if they don't throw NPE on `encryptorRunTime`, and now return `null` for unencrypted columns, then no problem.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java:
```diff
@@ -0,0 +1,317 @@
+            AesCipher.quickUpdatePageAAD(encryptorRunTime.getDataPageAAD(), pageOrdinal);
+          }
+          DataPageHeader headerV1 = pageHeader.data_page_header;
+          pageLoad = processPayload(reader, pageHeader.getCompressed_page_size(), encryptorRunTime.getDataEncryptor(), encryptorRunTime.getDataPageAAD(), encrypt);
```
should be handled differently if `encrypt` is false

shangxinli replied:
Same as above.

ggershinsky commented:
went over the code, adding a few more comments. Overall looks good; once the remaining points are handled, I'll be ok with the PR.

gszadovszky reviewed (changes requested):
I only have one minor comment. Meanwhile, this PR is conflicting with master. Please rebase.

shangxinli commented:
Just addressed your feedback. For the conflict, I tried to merge from upstream but it always resulted in a lot more conflict. So I just simply revert my change that partially renamed AAD to pageHeaderAAD. I looked at the util.java file and I see there are a lot of other places all using AAD instead of a more specific AAD. We can have another PR  to fix them all.

ggershinsky commented:
> So I just simply revert my change that partially renamed AAD to pageHeaderAAD. I looked at the util.java file and I see there are a lot of other places all using AAD instead of a more specific AAD. We can have another PR to fix them all.

I think this is fine. My earlier comment was addressed to code where it was not clear what kind of AADs are calculated/used. This is already fixed, thanks. As for the Util.java - it is very clear about the module kinds, with functions like writePageHeader etr. So its ok to leave the parameter name as is.

ggershinsky commented:
@shangxinli looks good, I think we're down to the 2 remaining points (related to a crypto-config flag) 
https://github.com/apache/parquet-mr/pull/928/files#r719567036
https://github.com/apache/parquet-mr/pull/928/files#r719567391

shangxinli commented:
Thank @ggershinsky and @gszadovszky for spending time on it!
